### PR TITLE
Allow prerender for all showcase routes and support for SSR/SSG

### LINKED ISSUE #153: Demo page SPA is not working as expected
This is described here basically:
https://github.com/isaacs/github/issues/408

Steps to reporduce:
1. Go to https://maplibre.org/ngx-maplibre-gl/
2. Click on examples
3. Click on any example
4. Refresh the browser
--> You'll get 404

This is because a SPA page assumes the server will return the index.html file for every unknown route, which doesn't happen in github as is explained in the issue above. 

The proposed solution there is to create redirects in cloudflare or use the instructions described here:
https://github.com/rafgraph/spa-github-pages

@birkskyum let me know if you would like to try out the CDN redirect.

### PR DESCRIPTION
**Pull request by birkskyum**
New angular applications will all use `application` builder (esbuild), but they don't migrate existing applications from `browser` builder (webpack), so I did that here following the first part of this [migration guide](https://angular.io/guide/esbuild#using-the-application-builder) on angular.json settings.

I also ran the `ng update @angular/cli --name=use-application-builder` on the showcase project as suggested in [here](https://github.com/angular/angular/issues/54965), and it [changed the tsconfig.json](https://github.com/maplibre/ngx-maplibre-gl/pull/156/commits/45f041a2fc8305cb71f4760b7470b5c1789983b2) a bit.

This keeps the app as a SPA, so it's still the 404 workaround it's using.

While technically not necessary for migrating to the `application`, I bumped all the deps as well, to make sure we don't hit some errors that are already patched. It was frictionless since there were only path/minors, except for typescript-eslint which introduced node 18+ requirement which align with the existing engine requirement in package.json.

Related to:
- #153

Fixes #153

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "ngx-maplibre-gl-srcs",
-  "version": "17.4.2",
+  "version": "^17.4.2",
```
Comment (HarelM) [inline package.json:3]: This is the version of the library, why setting it to "^"?

**Comment by HarelM** [comment]
Looks good in general, thanks!!

File: package.json (hunk 1 above)
Comment (birkskyum) [reply package.json:3]: oh shoot, good catch

**Review by HarelM — APPROVED** [review]

**Comment by HarelM** [comment]
Merging this will publish the docs and will result in bringing back the 404 bug: 
- #153

 Do you think a temporary solution should be added as described in the [spa-github-pages](https://github.com/rafgraph/spa-github-pages) repo until a proper fix is made?
Adding the above solution is a 5 minutes change to the CI and index.html file.

**Comment by birkskyum** [comment]
Oh - How does this bring bag a the 404 more than the existing SPA setup?

**Comment by HarelM** [comment]
Because I only updated the gh-pages branch in order to test that the fix is working and haven't made any changes to the code to incorporate this change - so redploying the gh-pages will remove my hot fix production code.
I'm not proud of what I did :-)

**Comment by HarelM** [comment]
Your code changes are not to blame.

**Comment by birkskyum** [comment]
Aha, I think that fix did the trick, so maybe we can just apply it here as well and it'll likely work for now. It is only temporary because I actually did manage to get most of the page SSR/SSG support yesterday through the afterNextRender, I'm just fighting 1-2 of the examples still, so that will properly fix the issue.

**Comment by birkskyum** [comment]
I'll make a PR to the gh-pages with the fix after merge.

**Comment by birkskyum** [comment]
Okay, I'll make a PR against this PR.

**Comment by birkskyum** [comment]
@HarelM , can you outline what the current issues are? Some routes not working, but it's somewhat unclear which?

**Comment by HarelM** [comment]
I just merged and update my dev env, I'll check locally and update here.

**Comment by HarelM** [comment]
So here's the current status, as far as I can tell.
The changes in the life cycle (ngOnInit -> afterFirstRender) is causing one test to fail. I didn't manage to understand how to fix it properly, Gimini and Google search didn't help me here.

When building using `npm run build:showcase:local` there are folders and files in the `dist/showcase/browser` folder according to what is written in the `route.txt` file as expected.

I guess the current way forward is to add a single route, see that it finish compilation.
While working on this, I think it might be useful to start a static file server instead of ng serve for both the tests to run in CI and to actually simulate how the github pages work.

I don't think a lot of work is still needed thanks to @marcjulian great work.

**Comment by HarelM** [comment]
Another route which is causing issue is `/demo/marker-alignment`.
I think these are the only two that needs a fix from what I can tell.
I've commented them out in the e2e test to see if it passes now.

**Comment by birkskyum** [comment]
Why does the runner use macos-latest? Isn't it much faster to access a linux instance?

**Comment by HarelM** [comment]
I don't remember, but it might have been some issues in linux and cypress tests, or maybe just xvfb extra complexity that I wanted to avoid.

**Comment by HarelM** [comment]
OK, All the e2e tests are passing.
All the routes render.
Still todo:
1. Find a way to unit test with afterNextRender
2. Search is not working anymore due to the fact that all the components are using the same selector `showcase-demo`, I'm not sure if this is related to the change in the build or not, but it seems to create an issue.

Other than that, I think this is the final stretch.

**Comment by birkskyum** [comment]
This search? Seems to work okay on the production built on a static server
Screenshot 2024-04-06 at 23 05 14

**Comment by HarelM** [comment]
When I run `npm run start` and try to search I see a console error. 
What method are you using to start?

**Comment by birkskyum** [comment]
I just ran `npm run build:showcase:local`, and launched the dist/showcase/browser/ as root of a static server

**Comment by HarelM** [comment]
Yup, that's the error related to the search and the fact that all components have the same selector. 
In any case, I'm off to to bed...

**Comment by HarelM** [comment]
I'm also not 100% sure all components are SSR ready, image source component for example caused some issues to the e2e tests and I solved it in the containing component and not in the image source component code... 
So feel free to review the changes and suggest fixes.

**Comment by marcjulian** [comment]
Search should work now also in dev mode. The clear button had `mat-button` and `mat-icon-button` directive applied causing the issue `NG0300: Multiple components match node with tagname button: _MatButton and _MatIconButton`

**Comment by marcjulian** [comment]
> I'm also not 100% sure all components are SSR ready, image source component for example caused some issues to the e2e tests and I solved it in the containing component and not in the image source component code... So feel free to review the changes and suggest fixes.

I think `ImageSourceComponent` is SSR ready, because `this.mapService.mapLoaded$` only fires on `MapService.setup` which is only executed in the browser because `MapComponent` uses `afterNextRender` to setup the map.

**Comment by HarelM** [comment]
@marcjulian any chance you can add a e2e test to the search for searching and clearing so that we'd know if it breaks in the future?
Thanks for taking care of this!
The only item left is the ability to unit test the afterFirstRender stuff, I don't think it's gating for this PR though.

**Review by HarelM — APPROVED** [review]

**Comment by marcjulian** [comment]
I couldn't find anything specific for testing `afterNextRender` with karma/jasmin. 
I can have a look of setting up an e2e for the search, I can add a new PR for it

**Comment by marcjulian** [comment]
LGTM too! I can test the next release, or dev release, in an SSR app to see how it behaves.

**Comment by HarelM** [comment]
Ok, once CI passes I'll merge this and release a new version, hopefully this will solve the linked issue.

**Comment by birkskyum** [comment]
Lets try that.

**Comment by HarelM** [comment]
Underway here: #159

**Pull request merged**